src/main/scala/higherkindness/rules_scala/workers/common/AnnexScalaInstance.scala

```diff
@@ -1,30 +1,74 @@
+  // and it seems to work well, so I'm sticking with it.
+  // For reference: https://github.com/sbt/zinc/blob/1.5.x/internal/zinc-classpath/src/main/scala/sbt/internal/inc/ScalaInstance.scala
+  override val compilerJars: Array[File] = allJars
+
+  override val libraryJars: Array[File] = allJars
+    .filter(f => new URLClassLoader(Array(f.toURI.toURL)).findResource("library.properties") != null)
+
+  override val otherJars: Array[File] = allJars.diff(compilerJars ++ libraryJars)
+
+  override val actualVersion: String = {
+    val stream = new URLClassLoader(
+      compilerJars.map(_.toURI.toURL),
+      null
+    ).getResourceAsStream("compiler.properties")
 
-  override def version: String = actualVersion
-  override lazy val actualVersion: String = {
-    val stream = loader.getResourceAsStream("compiler.properties")
     try {
       val props = new Properties
       props.load(stream)
       props.getProperty("version.number")
     } finally stream.close()
   }
+  override val version: String = actualVersion
 
-  override def compilerJar: File = null
-  override lazy val libraryJars: Array[File] = allJars
-    .filter(f => new URLClassLoader(Array(f.toURI.toURL)).findResource("library.properties") != null)
+  override val loaderLibraryOnly: ClassLoader =
+    new URLClassLoader(
+      libraryJars.map(_.toURI.toURL),
+      // The Scala 3 compiler jar has a dependency on the org.scala-sbt:compiler-interface jar.
+      // Our ZincRunner code also has a dependency on the compiler-interface.
+      //
+      // This means we will end up in a situation with classes from the compiler-interface
+      // being loaded by the classloader for the jvm that launched ZincRunner and also
+      // the classloader that we're creating here for the Scala 3 compiler.
+      //
+      // This is a problem.
+      //
+      // An instance of a class in the JVM is defined by its fully qualified name
+      // and classloader. Because the classloaders are different, we will get a
+      // java.lang.LinkageError because the classloader for the ZincRunner will have already
+      // loaded some compiler-interface classes that get used by the Scala 3 compiler during
+      // its compilation run.
+      //
+      // To avoid this we are setting the parent of the Scala 3 compiler classloader, which
+      // we are creating here, to the classloader for the jvm that launched the ZincRunner.
+      // That avoids the LinkageError because URLClassLoaders go parent first, so it finds
+      // the compiler-interface classes from the correct classloader.
+      //
+      // Problem is, doing this creates a potential leak: the parent classloader will be checked
+      // for all classes used during the compilation. Meaning, any classes used in the
+      // ZincRunner could end up leaking out into the Scala 3 compiler. So far this hasn't
+      // caused any issues, but maybe it will in the future.
```

Comment:
Would it be safer to go the other way around and make this classloader the parent for the ZincRunner's classloader? It seems classes leaking into the Scala compiler would be more problematic than into the ZincRunner

Reply:
The ZincRunner's classloader is the classloader that happens when we launch the JVM. I am ignorant of whether you can/how you can control that classloader or do what you are suggesting. Not saying it's not possible, just that I don't know how. Essentially, I think the answer to your question is yes, but I don't know how to do it or whether it is possible. Especially considering this classloader is not created until well after the program has started executing.

I do think revisiting this in the future would be a good idea, but so far I haven't run into any issues with the leak.